### ISSUE TITLE
Immediate mode widget API

### ISSUE THREAD

**Issue by aclysma** [aclysma/skulpin#30]
This is probably out of scope for this project, but it could be useful to have an API like imgui that uses Skia for drawing. This would be a good complement to this project.

**Comment by Kethku** [comment]
One difficulty here is that there is no way to have a persistent image on the screen. The renderer as I understand it requires drawing the entire image every frame which means battery performance will always be pretty bad. Some kind of persistent image which lets you redraw only what is necessary would improve this significantly. Unfortunately I don't know what is involved in doing this...

**Comment by aclysma** [comment]
This is definitely worth some thought, and I think there’s even a TODO in the code about this. My experience has been with games where it’s not as much a concern because usually the image is totally different. I’m curious to know how chrome and flutter approach this since they are using the same stack (skia/vulkan) and have roughly the same problem to solve. 

Vulkan swapchains use a set of images and cycle through them. My code clears the image before handing it to you, because what you get back otherwise would be a couple frames old, and dependent on hardware and timing. This seemed like it would be confusing and unhelpful, and the clear is a sensible default for many cases. 

My first thought is that maybe it would make sense for you to create your own skia canvases so that you could cache parts of the screen. Then you could only re-render the parts of the screen you need and the per-frame operation is just a buffer copy. For example, the contents of a scrolling text buffer might be generated in chunks and reused as you scroll up and down. Of course you could keep your own full-screen buffer of the previous frame. I was considering doing that, but a clear seemed like a better approach for the time being. You may need to look at the renderer code to see examples of how to create new vulkan-backed canvases. I don’t think it would be difficult, but there may need to be something new in the API to get what you need to do it. 

It might also be possible to only redraw when you need it. Currently a skulpin app just redraws as fast as it can, possibly slowed by V-sync. Maybe it makes sense to add something to the interface that lets the end-user choose if a redraw is needed. You could look at the renderer_only example at the line window.request_redraw() - it’s a bit more boilerplate for you to use the library that way but you get more control.

Again, would be helpful to look at how others have solved this!

**Comment by Kethku** [comment]
I implemented what you suggested which definitely improved things, but there was a mild trade-off. I got less cpu usage in exchange for more gpu usage.

```Rust
let mut context = gpu_canvas.gpu_context().unwrap();
let budgeted = Budgeted::YES;
let image_info = gpu_canvas.image_info();
let surface_origin = SurfaceOrigin::TopLeft;
let mut surface = Surface::new_render_target(&mut context, budgeted, &image_info, None, surface_origin, None, None).expect("Could not create surface");
let mut canvas = surface.canvas();
coordinate_system_helper.use_logical_coordinates(&mut canvas);

if let Some(image) = self.image.as_ref() {
    canvas.draw_image(image, (0, 0), Some(&self.paint));
} else {
    canvas.clear(default_colors.background.clone().unwrap().to_color());
}

// Do optional expensive canvas drawing

self.image = Some(surface.image_snapshot());
coordinate_system_helper.use_physical_coordinates(gpu_canvas);
gpu_canvas.draw_image(self.image.as_ref().unwrap(), (0, 0), Some(&self.paint));
```

This lets me cache the image and only draw updated regions pretty well, but the gpu works a bit harder. I think for now this unblocks my over arching goals, but I still wish there was a render mode which didn't require redrawing the stored image every frame. Especially on large monitors this results in a non-trivial operation for the entire window and I think we can do better.

In terms of rendering to a surface, no further api is needed as the canvas exposes the necessary gpu context.

**Comment by aclysma** [comment]
Have you tried skipping the `window.request_redraw()` call?

You can also set the control_flow out parameter to something other than poll. As a quick test I tried setting it like this to essentially drop to 5fps..

`*control_flow = winit::event_loop::ControlFlow::WaitUntil(std::time::Instant::now() + std::time::Duration::from_secs_f32(0.2));`

On MacOS this alone cut power usage by 80%. After this change it no longer showed up in the "apps using significant energy" list. The nice thing about this is that `Wait` and `WaitUntil` will still deliver any input events as they happen, so you can still immediately respond to input when appropriate.

**Comment by aclysma** [comment]
I'm going to close this because I think it's out of scope for this project. (Even though I still think it would be really neat!)
